File: src/compiler/nullness.ts
// Reading jspecify nullness annotations (@Nullable / @NonNull / @NullMarked /
// @NullUnmarked, https://jspecify.dev/docs/spec/) off a declaration, and deciding
// whether a *target* position (a parameter, return, field or local) is non-null.
// The checker uses these to warn when a possibly-null value reaches a non-null
// position. This is purely syntactic: declared nullness is read, never narrowed
// (no flow analysis), and generic/array-component nullness is out of scope.

import {
  type Annotation,
  type FieldDeclaration,
  type LocalVariableDeclarationStatement,
  type MethodDeclaration,
  type ModifierLike,
  type Node,
  type NodeArray,
  type Parameter,
  type RecordComponent,
  type SourceFile,
  SyntaxKind,
} from "./types.ts";
import { entityNameToString } from "./utilities.ts";

/** The three states a target position can be in; "unknown" stays silent. */
export type Nullness = "nonNull" | "nullable" | "unknown";

/** The nullness config (FQDN lists), as it comes from cappu.json. */
export interface NullnessOptions {
  readonly enabled: boolean;
  readonly nullableAnnotations: readonly string[];
  readonly nonNullAnnotations: readonly string[];
  readonly nullMarkedAnnotations: readonly string[];
  readonly nullUnmarkedAnnotations: readonly string[];
}

/** Annotation simple-name sets the checker matches against (resolved once). */
export interface NullnessAnnotations {
  readonly nullable: ReadonlySet<string>;
  readonly nonNull: ReadonlySet<string>;
  readonly nullMarked: ReadonlySet<string>;
  readonly nullUnmarked: ReadonlySet<string>;
}

// Match by simple name so both `@Nullable` and `@org.jspecify.annotations.Nullable`
// hit the same configured entry (the same trick readDeprecation uses).
const simpleName = (qualified: string): string => qualified.replace(/^.*\./, "");

export function resolveNullnessAnnotations(options: NullnessOptions): NullnessAnnotations {
  const set = (xs: readonly string[]): ReadonlySet<string> => new Set(xs.map(simpleName));
  return {
    nullable: set(options.nullableAnnotations),
    nonNull: set(options.nonNullAnnotations),
    nullMarked: set(options.nullMarkedAnnotations),
    nullUnmarked: set(options.nullUnmarkedAnnotations),
  };
}

// A declaration's annotations live on `.modifiers` (types, methods, params,
// fields, locals) or on `.annotations` (package/module declarations).
function annotationsOf(node: Node): readonly Node[] {
  const n = node as { modifiers?: NodeArray<ModifierLike>; annotations?: NodeArray<Annotation> };
  return n.modifiers ?? n.annotations ?? [];
}

function hasAnnotation(node: Node | undefined, names: ReadonlySet<string>): boolean {
  if (!node) return false;
  for (const m of annotationsOf(node)) {
    if (m.kind !== SyntaxKind.Annotation) continue;
    if (names.has(simpleName(entityNameToString((m as Annotation).typeName)))) return true;
  }
  return false;
}

// The node carrying the modifiers + type for a symbol's declaration. A field or
// local's annotation sits on the enclosing declaration, not the VariableDeclarator.
export function carrierOf(decl: Node | undefined): Node | undefined {
  if (decl?.kind === SyntaxKind.VariableDeclarator) return decl.parent;
  return decl;
}

function typeNodeOf(carrier: Node): Node | undefined {
  switch (carrier.kind) {
    case SyntaxKind.MethodDeclaration:
      return (carrier as MethodDeclaration).returnType;
    case SyntaxKind.Parameter:
      return (carrier as Parameter).type;
    case SyntaxKind.FieldDeclaration:
      return (carrier as FieldDeclaration).type;
    case SyntaxKind.LocalVariableDeclarationStatement:
      return (carrier as LocalVariableDeclarationStatement).type;
    case SyntaxKind.RecordComponent:
      return (carrier as RecordComponent).type;
    default:
      return undefined;
  }
}

// Only reference types carry nullness; a primitive (or an unresolved `var`) never
// does. Arrays are reference types (the variable itself, not its elements).
function isReferenceType(type: Node | undefined): boolean {
  return type?.kind === SyntaxKind.TypeReference || type?.kind === SyntaxKind.ArrayType;
}

// Is `node` inside a @NullMarked scope? The nearest enclosing @NullMarked /
// @NullUnmarked on the declaration, an enclosing type, or this file's package
// declaration wins. Cross-file package-info.java is not consulted.
function isNullMarked(node: Node, a: NullnessAnnotations): boolean {
  for (let n: Node | undefined = node; n; n = n.parent) {
    if (hasAnnotation(n, a.nullUnmarked)) return false;
    if (hasAnnotation(n, a.nullMarked)) return true;
    if (n.kind === SyntaxKind.SourceFile) {
      const pkg = (n as SourceFile).packageDeclaration;
      if (hasAnnotation(pkg, a.nullUnmarked)) return false;
      if (hasAnnotation(pkg, a.nullMarked)) return true;
    }
  }
  return false;
}

/**
 * The declared nullness of a target position (the carrier node holding the
 * modifiers + type). Explicit @Nullable/@NonNull wins; otherwise a reference type
 * in a @NullMarked scope is non-null; everything else is unknown (silent).
 */
export function targetNullness(carrier: Node | undefined, a: NullnessAnnotations): Nullness {
  if (!carrier) return "unknown";
  if (hasAnnotation(carrier, a.nullable)) return "nullable";
  if (hasAnnotation(carrier, a.nonNull)) return "nonNull";
  if (!isReferenceType(typeNodeOf(carrier))) return "unknown";
  return isNullMarked(carrier, a) ? "nonNull" : "unknown";
}

File: togo/internal/compiler/nullness.go
package compiler

// Port of src/compiler/nullness.ts: reading jspecify nullness annotations
// (@Nullable / @NonNull / @NullMarked / @NullUnmarked, https://jspecify.dev/docs/spec/)
// off declarations and type nodes. The checker turns these into a nullness facet on
// the type model (see checker_types.go) and warns when a possibly-null value reaches
// a non-null position. Purely syntactic: declared nullness is read, never narrowed.

import (
	"strings"

	"github.com/nikeee/cappu/internal/config"
)

// Nullness is a position's nullness; "" means unknown and stays silent.
type Nullness string

const (
	NullnessNonNull  Nullness = "nonNull"
	NullnessNullable Nullness = "nullable"
)

// NullnessAnnotations are the annotation simple-name sets the checker matches
// against, resolved once from the config (FQDN) lists.
type NullnessAnnotations struct {
	nullable     map[string]bool
	nonNull      map[string]bool
	nullMarked   map[string]bool
	nullUnmarked map[string]bool
}

// Match by simple name so both `@Nullable` and `@org.jspecify.annotations.Nullable`
// hit the same configured entry (the same trick readDeprecation uses).
func simpleName(qualified string) string {
	if i := strings.LastIndex(qualified, "."); i >= 0 {
		return qualified[i+1:]
	}
	return qualified
}

func nameSet(xs []string) map[string]bool {
	s := make(map[string]bool, len(xs))
	for _, x := range xs {
		s[simpleName(x)] = true
	}
	return s
}

// resolveNullnessAnnotations turns the config lists into simple-name sets, or
// returns nil when nullness checking is disabled.
func resolveNullnessAnnotations(o *config.Nullness) *NullnessAnnotations {
	if o == nil || !o.Enabled {
		return nil
	}
	return &NullnessAnnotations{
		nullable:     nameSet(o.NullableAnnotations),
		nonNull:      nameSet(o.NonNullAnnotations),
		nullMarked:   nameSet(o.NullMarkedAnnotations),
		nullUnmarked: nameSet(o.NullUnmarkedAnnotations),
	}
}

// A declaration's annotations live on its modifiers (types, methods, params,
// fields, locals) or an `annotations` field (record components, package /
// module declarations, type-use nodes).
func nullnessAnnotationsOf(node *Node) []*Node {
	if node == nil {
		return nil
	}
	if mods := declModifiers(node); mods != nil {
		return mods.Nodes
	}
	switch node.Kind {
	case LocalVariableDeclarationStatement:
		return arrayNodes(node.AsLocalVariableDeclarationStatement().Modifiers)
	case RecordComponent:
		return arrayNodes(node.AsRecordComponent().Annotations)
	case PackageDeclaration:
		return arrayNodes(node.AsPackageDeclaration().Annotations)
	case TypeReference:
		return arrayNodes(node.AsTypeReference().Annotations)
	case PrimitiveType:
		return arrayNodes(node.AsPrimitiveType().Annotations)
	}
	return nil
}

// hasNullnessAnnotation reports whether node carries an annotation in names.
func hasNullnessAnnotation(node *Node, names map[string]bool) bool {
	if node == nil {
		return false
	}
	for _, m := range nullnessAnnotationsOf(node) {
		if m.Kind != Annotation {
			continue
		}
		if names[simpleName(entityNameToString(m.AsAnnotation().TypeName))] {
			return true
		}
	}
	return false
}

func nullnessFrom(node *Node, a *NullnessAnnotations) Nullness {
	if hasNullnessAnnotation(node, a.nullable) {
		return NullnessNullable
	}
	if hasNullnessAnnotation(node, a.nonNull) {
		return NullnessNonNull
	}
	return ""
}

// carrierOf returns the node carrying the modifiers for a symbol's declaration. A
// field or local's annotation sits on the enclosing declaration, not the declarator.
func carrierOf(decl *Node) *Node {
	if decl != nil && decl.Kind == VariableDeclarator {
		return decl.Parent
	}
	return decl
}

// readDeclaredNullness is the nullness declared by a declaration's own modifiers.
func readDeclaredNullness(decl *Node, a *NullnessAnnotations) Nullness {
	return nullnessFrom(carrierOf(decl), a)
}

// typeUseNullness is the nullness written as a type-use annotation on a type node.
func typeUseNullness(typeNode *Node, a *NullnessAnnotations) Nullness {
	return nullnessFrom(typeNode, a)
}

// isReferenceTypeNode reports whether a type node is a reference type (only those
// carry nullness; a primitive or `var` never does). Arrays are reference types.
func isReferenceTypeNode(t *Node) bool {
	return t != nil && (t.Kind == TypeReference || t.Kind == ArrayType)
}
